src/main_loop.rs

```diff
@@ -879,161 +879,196 @@ impl MainLoopHandler {
         Ok(())
     }
 
-    /// Perform peer discovery and (if necessary) reconnect to the peers listed
-    /// as CLI arguments. Peer discovery involves finding potential peers from
-    /// connected peers and attempts to establish connections with one of them.
+    /// If necessary, disconnect from peers.
+    ///
+    /// Never disconnects from peers listed as CLI arguments.
     ///
     /// Locking:
     ///   * acquires `global_state_lock` for read
-    async fn peer_discovery_and_reconnector(
-        &self,
-        main_loop_state: &mut MutableMainLoopState,
-    ) -> Result<()> {
-        let cli_args = self.global_state_lock.cli().clone();
-        let global_state = self.global_state_lock.lock_guard().await;
-
-        let connected_peers: Vec<PeerInfo> = global_state.net.peer_map.values().cloned().collect();
+    async fn prune_peers(&self) -> Result<()> {
+        // fetch all relevant info from global state, then release the lock
+        let cli_args = self.global_state_lock.cli();
+        let connected_peers = self
+            .global_state_lock
+            .lock_guard()
+            .await
+            .net
+            .peer_map
+            .values()
+            .cloned()
+            .collect_vec();
 
-        // Check if we are connected to too many peers
         let num_peers = connected_peers.len();
         let max_num_peers = cli_args.max_num_peers;
-        if num_peers > max_num_peers {
-            // If *all* peer connections were outgoing, then it's OK to exceed
-            // the max-peer count. But in that case we don't want to connect to
-            // more peers; stop execution here.
-            if connected_peers.iter().all(|x| x.connection_is_outbound()) {
-                return Ok(());
-            }
+        if num_peers <= max_num_peers {
+            return Ok(());
+        }
 
-            // This would indicate a race-condition on the peer map field in the state which
-            // we unfortunately cannot exclude. So we just disconnect from a peer that the user
-            // didn't request a connection to.
-            warn!(
-                "Max peer parameter is exceeded. max is {max_num_peers} but we are connected \
-                to {num_peers}. Attempting to fix."
-            );
+        // If *all* peer connections were outgoing, then it's OK to exceed
+        // the max-peer count. But in that case we don't want to connect to
+        // more peers; stop execution here.
```

Comment:
I think this comment is deprecated, as this function cannot make new connections anymore.